File: apps/Image-to-3D/Android/app/src/main/java/ai/zetic/demo/imageto3d/DepthTo3D.kt
package ai.zetic.demo.imageto3d

import android.graphics.Bitmap
import kotlin.math.tan

/** Geometry buffers for one reconstruction — mirrors iOS DepthTo3D exactly. */
class MeshData(
    val positions: FloatArray,       // xyz per vertex
    val uvs: FloatArray,             // uv per vertex
    val triangleIndices: IntArray,
    val pointIndices: IntArray,      // point-cloud subset (edge-band dropped, 2× thinned)
    val pointColors: FloatArray,     // rgb per vertex
    val center: FloatArray,          // xyz
    val radius: Float,
)

/**
 * Converts a relative inverse-depth map into a planar-relief mesh:
 * x/y fixed on a reference plane (photo keeps its shape), z displaced by
 * smoothed disparity; quads across RAW-disparity discontinuities are culled,
 * the cut eroded by one quad, and the mask despeckled.
 */
object DepthTo3D {
    fun build(depth: DepthMap, texture: Bitmap): MeshData {
        val size = depth.width
        val stride = AppConfig.MESH_STRIDE
        val grid = (size - 1) / stride + 1

        val focal = size / 2f / tan(Math.toRadians(AppConfig.FOV_Y_DEGREES / 2.0)).toFloat()
        val center = (size - 1) / 2f
        val unitsPerPixel = AppConfig.PLANE_Z / focal
        val relief = AppConfig.RELIEF_DEPTH * size * unitsPerPixel

        // Raw disparity on the vertex grid; smoothed copy for geometry only.
        val raw = FloatArray(grid * grid)
        for (gy in 0 until grid) {
            val v = gy * stride
            for (gx in 0 until grid) {
                raw[gy * grid + gx] = depth.normalized(v * size + gx * stride)
            }
        }
        val disparity = gaussianSmoothed(raw, grid)

        val positions = FloatArray(grid * grid * 3)
        val uvs = FloatArray(grid * grid * 2)
        for (gy in 0 until grid) {
            val v = gy * stride
            for (gx in 0 until grid) {
                val u = gx * stride
                val i = gy * grid + gx
                val d = disparity[i]
                positions[i * 3] = (u - center) * unitsPerPixel
                positions[i * 3 + 1] = -(v - center) * unitsPerPixel
                positions[i * 3 + 2] = -(AppConfig.PLANE_Z + (0.5f - d) * relief)
                uvs[i * 2] = u / (size - 1f)
                uvs[i * 2 + 1] = v / (size - 1f)
            }
        }

        // Quad validity from RAW disparity, then erode + despeckle.
        val qside = grid - 1
        var valid = BooleanArray(qside * qside)
        val threshold = AppConfig.EDGE_THRESHOLD
        for (gy in 0 until qside) {
            for (gx in 0 until qside) {
                val i00 = gy * grid + gx
                val d0 = raw[i00]
                val d1 = raw[i00 + 1]
                val d2 = raw[i00 + grid]
                val d3 = raw[i00 + grid + 1]
                val hi = maxOf(d0, d1, d2, d3)
                val lo = minOf(d0, d1, d2, d3)
                valid[gy * qside + gx] = (hi - lo) <= threshold
            }
        }
        valid = erodedValid(valid, qside)
        valid = cleaned(valid, qside)

        val indices = ArrayList<Int>(qside * qside * 6)
        for (gy in 0 until qside) {
            for (gx in 0 until qside) {
                if (!valid[gy * qside + gx]) continue
                val i00 = gy * grid + gx
                val i10 = i00 + 1
                val i01 = i00 + grid
                val i11 = i01 + 1
                indices.add(i00); indices.add(i01); indices.add(i10)
                indices.add(i10); indices.add(i01); indices.add(i11)
            }
        }

        // Point cloud: 2× thinned; drop vertices whose every adjacent quad is culled.
        val pointIndices = ArrayList<Int>(grid * grid / 4)
        for (gy in 0 until grid step 2) {
            for (gx in 0 until grid step 2) {
                var touched = false
                for ((qy, qx) in arrayOf(gy - 1 to gx - 1, gy - 1 to gx, gy to gx - 1, gy to gx)) {
                    if (qy in 0 until qside && qx in 0 until qside && valid[qy * qside + qx]) {
                        touched = true
                        break
                    }
                }
                if (touched) pointIndices.add(gy * grid + gx)
            }
        }

        // Per-vertex colors sampled from the texture.
        val texPixels = IntArray(size * size)
        texture.getPixels(texPixels, 0, size, 0, 0, size, size)
        val colors = FloatArray(grid * grid * 3)
        for (gy in 0 until grid) {
            val v = gy * stride
            for (gx in 0 until grid) {
                val p = texPixels[v * size + gx * stride]
                val i = gy * grid + gx
                colors[i * 3] = ((p shr 16) and 0xFF) / 255f
                colors[i * 3 + 1] = ((p shr 8) and 0xFF) / 255f
                colors[i * 3 + 2] = (p and 0xFF) / 255f
            }
        }

        // Bounds over triangle-referenced vertices.
        var loX = Float.MAX_VALUE; var loY = Float.MAX_VALUE; var loZ = Float.MAX_VALUE
        var hiX = -Float.MAX_VALUE; var hiY = -Float.MAX_VALUE; var hiZ = -Float.MAX_VALUE
        val refs = if (indices.isEmpty()) (0 until grid * grid) else indices
        for (i in refs) {
            val x = positions[i * 3]; val y = positions[i * 3 + 1]; val z = positions[i * 3 + 2]
            if (x < loX) loX = x; if (x > hiX) hiX = x
            if (y < loY) loY = y; if (y > hiY) hiY = y
            if (z < loZ) loZ = z; if (z > hiZ) hiZ = z
        }
        val dx = hiX - loX; val dy = hiY - loY; val dz = hiZ - loZ
        val radius = (Math.sqrt((dx * dx + dy * dy + dz * dz).toDouble()) / 2).toFloat()

        return MeshData(
            positions = positions,
            uvs = uvs,
            triangleIndices = indices.toIntArray(),
            pointIndices = pointIndices.toIntArray(),
            pointColors = colors,
            center = floatArrayOf((loX + hiX) / 2, (loY + hiY) / 2, (loZ + hiZ) / 2),
            radius = maxOf(radius, 0.05f),
        )
    }

    /** Separable 5-tap Gaussian ([1,4,6,4,1]/16) over the vertex grid. */
    private fun gaussianSmoothed(input: FloatArray, grid: Int): FloatArray {
        val k = floatArrayOf(1 / 16f, 4 / 16f, 6 / 16f, 4 / 16f, 1 / 16f)
        val tmp = FloatArray(input.size)
        val out = FloatArray(input.size)
        for (gy in 0 until grid) {
            for (gx in 0 until grid) {
                var acc = 0f
                for (t in -2..2) {
                    val x = (gx + t).coerceIn(0, grid - 1)
                    acc += input[gy * grid + x] * k[t + 2]
                }
                tmp[gy * grid + gx] = acc
            }
        }
        for (gy in 0 until grid) {
            for (gx in 0 until grid) {
                var acc = 0f
                for (t in -2..2) {
                    val y = (gy + t).coerceIn(0, grid - 1)
                    acc += tmp[y * grid + gx] * k[t + 2]
                }
                out[gy * grid + gx] = acc
            }
        }
        return out
    }

    /** Valid only when the full 8-neighborhood is valid (dilates the cut). */
    private fun erodedValid(mask: BooleanArray, side: Int): BooleanArray {
        val out = mask.copyOf()
        for (y in 0 until side) {
            for (x in 0 until side) {
                if (!mask[y * side + x]) continue
                var keep = true
                for (dy in -1..1) {
                    for (dx in -1..1) {
                        val ny = (y + dy).coerceIn(0, side - 1)
                        val nx = (x + dx).coerceIn(0, side - 1)
                        if (!mask[ny * side + nx]) keep = false
                    }
                }
                out[y * side + x] = keep
            }
        }
        return out
    }

    /** Majority filter: flip quads whose 3×3 neighborhood vote is lopsided. */
    private fun cleaned(mask: BooleanArray, side: Int): BooleanArray {
        val out = mask.copyOf()
        for (y in 0 until side) {
            for (x in 0 until side) {
                var votes = 0
                var total = 0
                for (dy in -1..1) {
                    for (dx in -1..1) {
                        if (dx == 0 && dy == 0) continue
                        val ny = y + dy
                        val nx = x + dx
                        if (ny !in 0 until side || nx !in 0 until side) continue
                        total++
                        if (mask[ny * side + nx]) votes++
                    }
                }
                val idx = y * side + x
                if (mask[idx] && votes <= total / 4) out[idx] = false
                if (!mask[idx] && votes >= total - total / 4) out[idx] = true
            }
        }
        return out
    }
}

File: apps/Image-to-3D/iOS/ImageTo3D/Rendering/DepthTo3D.swift
import Foundation
import UIKit
import simd

/// Geometry buffers for one reconstruction: a textured triangle grid and the
/// matching colored point cloud (same vertex positions).
struct MeshData {
    var positions: [SIMD3<Float>]
    var uvs: [SIMD2<Float>]
    var triangleIndices: [UInt32]
    var pointIndices: [UInt32]   // point-cloud subset (edge-band vertices dropped)
    var pointColors: [Float]     // packed RGB triplets (3 floats per vertex)
    var boundsMin: SIMD3<Float>
    var boundsMax: SIMD3<Float>

    var center: SIMD3<Float> { (boundsMin + boundsMax) / 2 }
    var radius: Float { simd_length(boundsMax - boundsMin) / 2 }
}

/// Converts a relative inverse-depth map into a 3D grid mesh via pinhole
/// unprojection. The camera sits at the origin looking down -Z, +Y up.
enum DepthTo3D {
    static func build(depth: DepthMap, texture: UIImage) -> MeshData {
        let size = depth.width
        let stride = AppConfig.meshStride
        let grid = (size - 1) / stride + 1

        // Planar-relief coordinates: x/y are fixed on the reference plane
        // (world units per source pixel), z is displaced by disparity.
        let focal = Float(size) / 2 / tan(AppConfig.fovYDegrees * .pi / 180 / 2)
        let center = Float(size - 1) / 2
        let unitsPerPixel = AppConfig.planeZ / focal
        let relief = AppConfig.reliefDepth * Float(size) * unitsPerPixel

        // Sample normalized disparity onto the vertex grid. Geometry uses a
        // smoothed copy (pleasant relief); the discontinuity test uses the RAW
        // values — smoothing spreads the fg/bg jump over several pixels, which
        // lets stretched "skirt" quads slip under the threshold and show up as
        // a hairy dark fringe around silhouettes.
        var rawDisparity = [Float](repeating: 0, count: grid * grid)
        for gy in 0..<grid {
            let v = gy * stride
            for gx in 0..<grid {
                rawDisparity[gy * grid + gx] = depth.normalized(v * size + gx * stride)
            }
        }
        let disparity = gaussianSmoothed(rawDisparity, grid: grid)

        let textureRGB = sampleTexture(texture, gridSize: grid, imageStride: stride)

        var positions = [SIMD3<Float>]()
        var uvs = [SIMD2<Float>]()
        positions.reserveCapacity(grid * grid)
        uvs.reserveCapacity(grid * grid)

        for gy in 0..<grid {
            let v = gy * stride
            for gx in 0..<grid {
                let u = gx * stride
                let d = disparity[gy * grid + gx]
                // Image v grows downward, scene Y grows upward → flip Y.
                // d = 1 (closest) pushes toward the camera, d = 0 away.
                positions.append(SIMD3<Float>((Float(u) - center) * unitsPerPixel,
                                              -(Float(v) - center) * unitsPerPixel,
                                              -(AppConfig.planeZ + (0.5 - d) * relief)))
                uvs.append(SIMD2<Float>(Float(u) / Float(size - 1),
                                        Float(v) / Float(size - 1)))
            }
        }

        // Quad validity from RAW disparity: a quad spanning a depth
        // discontinuity is culled whole. The cut is then dilated one quad so
        // anti-aliased edge pixels (the dark outline smear) go with it, and
        // despeckled so isolated spikes/pinholes vanish.
        let qside = grid - 1
        var valid = [Bool](repeating: false, count: qside * qside)
        let threshold = AppConfig.edgeThreshold
        for gy in 0..<qside {
            for gx in 0..<qside {
                let i00 = gy * grid + gx
                let d0 = rawDisparity[i00]
                let d1 = rawDisparity[i00 + 1]
                let d2 = rawDisparity[i00 + grid]
                let d3 = rawDisparity[i00 + grid + 1]
                let hi = max(max(d0, d1), max(d2, d3))
                let lo = min(min(d0, d1), min(d2, d3))
                valid[gy * qside + gx] = (hi - lo) <= threshold
            }
        }
        valid = erodedValid(valid, side: qside)
        valid = cleaned(valid, side: qside)

        var indices = [UInt32]()
        indices.reserveCapacity(qside * qside * 6)
        for gy in 0..<qside {
            for gx in 0..<qside {
                guard valid[gy * qside + gx] else { continue }
                let i00 = UInt32(gy * grid + gx)
                let i10 = i00 + 1
                let i01 = i00 + UInt32(grid)
                let i11 = i01 + 1
                indices.append(contentsOf: [i00, i01, i10, i10, i01, i11])
            }
        }

        // Point cloud: skip vertices whose every adjacent quad is culled —
        // they sit inside the smeared discontinuity band and render as a
        // noisy halo around silhouettes. Thinned 2× (full grid is moiré soup).
        var pointIndices = [UInt32]()
        pointIndices.reserveCapacity(grid * grid / 4)
        for gy in Swift.stride(from: 0, to: grid, by: 2) {
            for gx in Swift.stride(from: 0, to: grid, by: 2) {
                var touched = false
                for (qy, qx) in [(gy - 1, gx - 1), (gy - 1, gx), (gy, gx - 1), (gy, gx)] {
                    if qy >= 0, qy < qside, qx >= 0, qx < qside, valid[qy * qside + qx] {
                        touched = true
                        break
                    }
                }
                if touched { pointIndices.append(UInt32(gy * grid + gx)) }
            }
        }

        // Bounds over vertices actually referenced by triangles, so stray
        // unreferenced grid corners don't inflate the camera framing.
        var lo = SIMD3<Float>(repeating: .greatestFiniteMagnitude)
        var hi = SIMD3<Float>(repeating: -.greatestFiniteMagnitude)
        if indices.isEmpty {
            for p in positions { lo = simd_min(lo, p); hi = simd_max(hi, p) }
        } else {
            for i in indices {
                let p = positions[Int(i)]
                lo = simd_min(lo, p)
                hi = simd_max(hi, p)
            }
        }

        return MeshData(positions: positions,
                        uvs: uvs,
                        triangleIndices: indices,
                        pointIndices: pointIndices,
                        pointColors: textureRGB,
                        boundsMin: lo,
                        boundsMax: hi)
    }

    /// Separable 5-tap Gaussian ([1,4,6,4,1]/16) over the vertex grid.
    private static func gaussianSmoothed(_ input: [Float], grid: Int) -> [Float] {
        let kernel: [Float] = [1 / 16.0, 4 / 16.0, 6 / 16.0, 4 / 16.0, 1 / 16.0]
        var tmp = [Float](repeating: 0, count: input.count)
        var out = [Float](repeating: 0, count: input.count)
        for gy in 0..<grid {
            for gx in 0..<grid {
                var acc: Float = 0
                for k in -2...2 {
                    let x = min(max(gx + k, 0), grid - 1)
                    acc += input[gy * grid + x] * kernel[k + 2]
                }
                tmp[gy * grid + gx] = acc
            }
        }
        for gy in 0..<grid {
            for gx in 0..<grid {
                var acc: Float = 0
                for k in -2...2 {
                    let y = min(max(gy + k, 0), grid - 1)
                    acc += tmp[y * grid + gx] * kernel[k + 2]
                }
                out[gy * grid + gx] = acc
            }
        }
        return out
    }

    /// Erodes the valid region by one quad (a quad stays valid only if its
    /// full 8-neighborhood is valid) — dilates the discontinuity cut so the
    /// anti-aliased silhouette pixels fall inside it.
    private static func erodedValid(_ mask: [Bool], side: Int) -> [Bool] {
        var out = mask
        for y in 0..<side {
            for x in 0..<side {
                guard mask[y * side + x] else { continue }
                var keep = true
                for dy in -1...1 {
                    for dx in -1...1 {
                        let ny = min(max(y + dy, 0), side - 1)
                        let nx = min(max(x + dx, 0), side - 1)
                        if !mask[ny * side + nx] { keep = false }
                    }
                }
                out[y * side + x] = keep
            }
        }
        return out
    }

    /// Majority filter on the quad mask: a quad flips to match its 3×3
    /// neighborhood when the vote is lopsided. Removes one-quad sawtooth
    /// spikes along silhouettes and pinholes inside surfaces while leaving
    /// genuine edges alone.
    private static func cleaned(_ mask: [Bool], side: Int) -> [Bool] {
        var out = mask
        for y in 0..<side {
            for x in 0..<side {
                var votes = 0
                var total = 0
                for dy in -1...1 {
                    for dx in -1...1 {
                        let ny = y + dy
                        let nx = x + dx
                        guard ny >= 0, ny < side, nx >= 0, nx < side,
                              !(dx == 0 && dy == 0) else { continue }
                        total += 1
                        if mask[ny * side + nx] { votes += 1 }
                    }
                }
                let idx = y * side + x
                if mask[idx] && votes <= total / 4 { out[idx] = false }
                if !mask[idx] && votes >= total - total / 4 { out[idx] = true }
            }
        }
        return out
    }

    /// Samples the 518×518 texture at each grid vertex → packed RGB floats.
    private static func sampleTexture(_ texture: UIImage, gridSize: Int, imageStride: Int) -> [Float] {
        let size = AppConfig.inputSize
        var colors = [Float](repeating: 0.8, count: gridSize * gridSize * 3)
        guard let cgImage = texture.cgImage else { return colors }

        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue
            | CGBitmapInfo.byteOrder32Big.rawValue)
        var pixels = [UInt8](repeating: 0, count: size * size * 4)
        let drawn: Bool = pixels.withUnsafeMutableBufferPointer { ptr in
            guard let context = CGContext(data: ptr.baseAddress,
                                          width: size,
                                          height: size,
                                          bitsPerComponent: 8,
                                          bytesPerRow: size * 4,
                                          space: colorSpace,
                                          bitmapInfo: bitmapInfo.rawValue) else { return false }
            context.draw(cgImage, in: CGRect(x: 0, y: 0, width: size, height: size))
            return true
        }
        guard drawn else { return colors }

        for gy in 0..<gridSize {
            let v = gy * imageStride
            for gx in 0..<gridSize {
                let u = gx * imageStride
                let src = (v * size + u) * 4
                let dst = (gy * gridSize + gx) * 3
                colors[dst] = Float(pixels[src]) / 255.0
                colors[dst + 1] = Float(pixels[src + 1]) / 255.0
                colors[dst + 2] = Float(pixels[src + 2]) / 255.0
            }
        }
        return colors
    }
}
